File: 4290-valid-elements-in-an-array/valid-elements-in-an-array.py
class Solution(object):
    def findValidElements(self, nums):
        n=len(nums)
        if n==0:
            return []
        stack=[]
        winners=set()
        left_max=float('-inf')
        for i in range (n):
            current=nums[i]
            if current > left_max :
                winners.add(i)
                left_max=current
            while stack and current >= nums[stack[-1]]:
                stack.pop()
            stack.append(i)
        winners.update(stack)
        result =[]
        for i in range(n):
            if i in winners :
                result.append(nums[i])
        return result


        
        

File: 4290-valid-elements-in-an-array/valid-elements-in-an-array.cpp
class Solution {
public:
    vector<int> findValidElements(vector<int>& nums) {
        int n = nums.size();
        int left_greatest = INT_MIN;
        int right_greatest = INT_MIN;
        vector<bool> valid(n, false);
        vector<int> result;

        // left pass
        for(int i=0; i<n; i++){
            if(nums[i]>left_greatest){
                valid[i] = true;
                left_greatest = nums[i];
            }
        }

        // right pass
        for(int i=n-1; i>=0; i--){
            if(nums[i]>right_greatest){
                valid[i] = true;
                right_greatest = nums[i];
            }
        }

        for(int i=0; i<n; i++){
            if(valid[i]) result.push_back(nums[i]);
        }

        return result;
    }
};